File: bindings/java/src/test/java/org/wickra/exchange/ExchangeTest.java
package org.wickra.exchange;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class ExchangeTest {

    @Test
    void versionIsExposed() {
        assertFalse(Exchange.version().isEmpty());
    }

    @Test
    void paperMarketBuyFills() {
        try (Exchange ex = Exchange.paper(Map.of("USDT", 100_000.0), 1.0, 5.0, 10.0)) {
            assertEquals("paper", ex.name());
            ex.setPrice("BTC/USDT", 20_000.0);

            Exchange.OrderInfo order = ex.placeMarket("BTC/USDT", Exchange.Side.BUY, 1.0);
            assertTrue(order.isFilled());
            // 10 bps slippage on a buy: 20000 * 1.001 = 20020.
            assertTrue(Math.abs(order.averagePrice() - 20_020.0) < 1e-6);

            assertTrue(Math.abs(ex.balance("BTC") - 1.0) < 1e-9);
            assertTrue(Math.abs(ex.balance("USDT") - (100_000.0 - 20_020.0 - 10.01)) < 1e-6);
        }
    }

    @Test
    void restingLimitAndCancel() {
        try (Exchange ex = Exchange.paper(Map.of("USDT", 100_000.0), 0.0, 0.0, 0.0)) {
            ex.setPrice("BTC/USDT", 20_000.0);
            Exchange.OrderInfo resting = ex.placeLimit("BTC/USDT", Exchange.Side.BUY, 1.0, 19_000.0);
            assertEquals(Exchange.Status.NEW, resting.status());
            List<Exchange.Event> events = ex.poll(16);
            assertTrue(events.stream().anyMatch(e -> e.kind() == Exchange.Kind.ORDER_UPDATE));
            ex.cancel("BTC/USDT", resting.id());
            assertTrue(Math.abs(ex.balance("USDT") - 100_000.0) < 1e-9);
        }
    }

    @Test
    void replayParity() {
        double[] tape = {100.0, 101.0, 102.0, 110.0, 112.0};
        try (Exchange ex = Exchange.replayTrades("BTC/USDT", tape, Map.of("USDT", 100_000.0), 0.0, 0.0, 0.0)) {
            assertEquals("replay", ex.name());

            double[] window = new double[3];
            int seen = 0;
            boolean bought = false;

            while (true) {
                List<Exchange.Event> events = ex.poll(16);
                if (events.isEmpty()) {
                    break;
                }
                for (Exchange.Event ev : events) {
                    if (!ev.isTrade()) {
                        continue;
                    }
                    double price = ev.price();
                    window[seen % 3] = price;
                    seen++;
                    if (seen >= 3) {
                        double mean = (window[0] + window[1] + window[2]) / 3.0;
                        if (!bought && price > mean) {
                            Exchange.OrderInfo order = ex.placeMarket("BTC/USDT", Exchange.Side.BUY, 1.0);
                            assertTrue(order.isFilled());
                            bought = true;
                        }
                    }
                }
            }

            assertTrue(bought);
            assertTrue(Math.abs(ex.balance("BTC") - 1.0) < 1e-9);
        }
    }
}

File: bindings/node/__tests__/exchange.test.js
"use strict";

const test = require("node:test");
const assert = require("node:assert");
const { Exchange, Credentials, OrderRequest, version } = require("../index.js");

test("module surface", () => {
  assert.strictEqual(typeof version(), "string");
  assert.strictEqual(typeof Exchange, "function");
  assert.strictEqual(typeof Credentials, "function");
  assert.strictEqual(typeof OrderRequest, "function");
});

test("credentials construct", () => {
  new Credentials("key", "secret");
  new Credentials("key", "secret", "passphrase");
  new Credentials("key", "secret", null, "-----BEGIN-----");
});

test("paper market buy fills with slippage and fee", () => {
  const ex = Exchange.paper({ USDT: 100000 }, 1, 5, 10);
  assert.strictEqual(ex.name(), "paper");
  ex.setPrice("BTC/USDT", 20000);

  const order = ex.placeOrder(OrderRequest.marketBuy("BTC/USDT", 1));
  assert.strictEqual(order.status, "filled");
  assert.strictEqual(order.side, "buy");
  // 10 bps slippage on a buy: 20000 * 1.001 = 20020.
  assert.ok(Math.abs(order.averagePrice - 20020) < 1e-6);

  const balances = ex.balances();
  assert.ok(Math.abs(balances.BTC - 1) < 1e-9);
  // 20020 notional + 5 bps fee (10.01) spent from USDT.
  assert.ok(Math.abs(balances.USDT - (100000 - 20020 - 10.01)) < 1e-6);
});

test("events and ticker", () => {
  const ex = Exchange.paper({ USDT: 50000 });
  ex.setPrice("BTC/USDT", 20000);
  ex.placeOrder(OrderRequest.marketBuy("BTC/USDT", 1));

  const kinds = ex.pollEvents().map((e) => e.kind);
  assert.ok(kinds.includes("order_update"));
  assert.ok(kinds.includes("balance_update"));

  const ticker = ex.ticker("BTC/USDT");
  assert.ok(Math.abs(ticker.last - 20000) < 1e-9);
});

test("resting limit and cancel", () => {
  const ex = Exchange.paper({ USDT: 100000 });
  ex.setPrice("BTC/USDT", 20000);
  const order = ex.placeOrder(OrderRequest.limitBuy("BTC/USDT", 1, 19000));
  assert.strictEqual(order.status, "new");
  ex.cancelOrder("BTC/USDT", order.id);
  assert.ok(Math.abs(ex.balances().USDT - 100000) < 1e-9);
});

test("bad market string throws", () => {
  const ex = Exchange.paper({ USDT: 1 });
  assert.throws(() => ex.setPrice("BTCUSDT", 1));
});
